### PR TITLE
[ONNX] Fix lstm reshape shape inference regression

### LINKED ISSUE #72399: PyTorch 1.10.2 failed to create workable onnx for sequential LSTMs
### 🐛 Describe the bug

The following code export runnable onnx until PyTorch 1.10 (tested and still work in 1.8 & 1.9), the exported onnx no longer work after that.

```
import torch
import torch.nn as nn

class testModel(nn.Module):
    def __init__(self):
        super(testModel, self).__init__()
        self.rnn1 = nn.LSTM(8, 8, bidirectional=True, batch_first=True)
        self.linear1 = nn.Linear(8 * 2, 8)

        self.rnn2 = nn.LSTM(8, 8, bidirectional=True, batch_first=True)
        self.linear2 = nn.Linear(8 * 2, 8)

    def forward(self, input):
        rnn_output1, _ = self.rnn1(input)
        linear_output1 = self.linear1(rnn_output1)

        rnn_output2, _ = self.rnn2(linear_output1)
        linear_output2 = self.linear2(rnn_output2)

        return linear_output2

net = testModel()
net.eval()
input = torch.zeros((1, 100, 8), dtype=torch.float32)

with torch.no_grad():
   output = net(input)


torch.onnx.export(
        net,
        input,
        'out.onnx',
        export_params=True,
        opset_version=13, 
        verbose=False,
        do_constant_folding=True,
        input_names = ['input'],
        output_names = ['output'],
        dynamic_axes={
                        'input' : {0 : 'batch_size',1:'w',2:'h'},
                        'output' : {0 : 'batch_size', 1:'w',2:'h'},
                }
        )

##### ONNXRuntime ####
import onnxruntime as rt
import numpy as np
sess = rt.InferenceSession('out.onnx')
result_1 = sess.run([], {'input': input.numpy()})[0] # Failed
``` 

Error Message from onnxruntime
```
[E:onnxruntime:, sequential_executor.cc:346 Execute] Non-zero status code returned while running LSTM node. Name:'LSTM_38' Status Message: Input initial_h must have shape {2,1,8}. Actual:{2,0,8}
```

After inspect the model with Netron, it seem like PyTorch produce invalid expand layers in the circle
out onnx-2

Workable onnx graph from PyTorch 1.9.0 (generated from above code)
out onnx-3

### Versions

pytorch 1.10.2
onnxruntime 1.10
onnx 1.10.2

cc @malfet @albanD

### PR DESCRIPTION
**Pull request by BowenBao**
Fixes #72399

### CODE REVIEW

**Comment by malfet** [comment]
@garymm this looks like a good candidate for GHF merge

**Comment by albanD** [comment]
Should this be cherry picked for 1.11 since it looks like a regression for this release?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 1]
```diff
@@ -1029,46 +931,48 @@ void ProcessReduceNode(Node* n) {
...
 
-  if (ConstantValueMap::HasShape(n->input(1)->debugName())) {
-    auto shape_vector_1 =
-        ConstantValueMap::GetShapeInto1DInt64Vector(n->input(1)->debugName());
-    if (shape_vector_1.has_value()) {
-      TORCH_INTERNAL_ASSERT(shape_vector_1.value().size() == 1);
-      UpdateRank(n->output(), shape_vector_1.value()[0]);
+  // Only shape of new shape is known, assign output rank.
+  if (ConstantValueMap::HasShape(shape_name)) {
+    auto static_shape_shape =
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:971]:
```suggestion
    auto output_rank =
```

**Review by garymm — APPROVED** [review]

**Comment by garymm** [comment]
> this looks like a good candidate for GHF merge

@malfet what does GHF merge mean?
Do you mean the pytorch merge bot?

**Pull request closed without merging**